File: Day05/part-01.js
let pages = [];
let allPages = new Set();
let positions = {};
let dependencies = {};
let updateIndex = 0;
let step = 0;
let maxSteps = 0;
let validUpdates = [];
let middlePages = [];
let sortingDone = false;

document
  .getElementById("fileInput")
  .addEventListener("change", handleFile, false);

function handleFile(event) {
  const file = event.target.files[0];
  const reader = new FileReader();
  reader.onload = function (event) {
    const text = event.target.result;
    processInput(text);
  };
  reader.readAsText(file);
}

function processInput(text) {
  const lines = text.trim().split("\n");
  const separatorIndex = lines.indexOf("");
  const rules = lines.slice(0, separatorIndex);
  const updates = lines
    .slice(separatorIndex + 1)
    .map((line) => line.split(",").map(Number));

  dependencies = buildDependencies(rules);
  validUpdates = updates.filter((update) => isCorrectOrder(update));

  // Add all pages from updates to the set
  updates.forEach((update) => update.forEach((page) => allPages.add(page)));

  if (validUpdates.length > 0) {
    updateIndex = 0;
    step = 0;
    sortingDone = false;
    setupCanvas(validUpdates[updateIndex]);
  } else {
    document.getElementById("result").innerText = "No valid updates found.";
  }
}

function buildDependencies(rules) {
  const deps = {};
  rules.forEach((rule) => {
    const [before, after] = rule.split("|").map(Number);
    if (!deps[after]) deps[after] = new Set();
    deps[after].add(before);
  });
  return deps;
}

function isCorrectOrder(update) {
  const positions = {};
  update.forEach((page, index) => (positions[page] = index));

  for (const [after, befores] of Object.entries(dependencies)) {
    if (positions.hasOwnProperty(after)) {
      for (const before of befores) {
        if (
          positions.hasOwnProperty(before) &&
          positions[before] > positions[after]
        ) {
          return false;
        }
      }
    }
  }
  return true;
}

function setupCanvas(update) {
  pages = [...allPages]; // Display all pages including those not in the current update
  maxSteps = update.length;
  positions = {};
  update.forEach((page, index) => {
    positions[page] = index;
  });
  updateCanvas();
}

function updateCanvas() {
  createCanvas(1500, 800).parent("canvasContainer");
  background(255);
  fill(0);
  textSize(24);
  textAlign(CENTER, CENTER);
  const gridSize = Math.ceil(Math.sqrt(pages.length));
  const cellSize = width / gridSize;

  pages.forEach((page, index) => {
    const x = (index % gridSize) * cellSize + cellSize / 2;
    const y = Math.floor(index / gridSize) * cellSize + cellSize / 2;
    if (positions.hasOwnProperty(page)) {
      fill(positions[page] === index ? "green" : "red");
    } else {
      fill("black"); // Pages not in the current update
    }
    text(page, x, y);
  });
}

function draw() {
  if (!sortingDone && validUpdates.length > 0) {
    const update = validUpdates[updateIndex];
    if (step < maxSteps - 1) {
      if (positions[update[step]] > positions[update[step + 1]]) {
        [positions[update[step]], positions[update[step + 1]]] = [
          positions[update[step + 1]],
          positions[update[step]],
        ];
        [update[step], update[step + 1]] = [update[step + 1], update[step]];
      }
      step++;
      updateCanvas();
    } else {
      step = 0;
      const middlePage = update[Math.floor(update.length / 2)];
      middlePages.push(middlePage);
      updateIndex++;
      if (updateIndex >= validUpdates.length) {
        sortingDone = true;
        const sumOfMiddlePages = middlePages.reduce((acc, val) => acc + val, 0);
        document.getElementById(
          "result"
        ).innerText = `Sorting completed. Sum of middle pages: ${sumOfMiddlePages}`;
      } else {
        setupCanvas(validUpdates[updateIndex]);
      }
    }
  }
}

File: Day05/part-01.py
def read_input(file_path):
    with open(file_path, 'r') as file:
        lines = file.read().strip().split('\n')
    separator_index = lines.index('')
    rules = lines[:separator_index]
    updates = [list(map(int, line.split(','))) for line in lines[separator_index + 1:]]
    return rules, updates

def build_dependencies(rules):
    dependencies = {}
    for rule in rules:
        before, after = map(int, rule.split('|'))
        if after not in dependencies:
            dependencies[after] = set()
        dependencies[after].add(before)
    return dependencies

def is_update_valid(update, dependencies):
    position = {page: idx for idx, page in enumerate(update)}

    for after, befores in dependencies.items():
        if after in position:
            for before in befores:
                if before in position and position[before] > position[after]:
                    return False
    return True

def main(file_path):
    rules, updates = read_input(file_path)
    dependencies = build_dependencies(rules)
    
    valid_updates = [update for update in updates if is_update_valid(update, dependencies)]
    middle_pages = [update[len(update) // 2] for update in valid_updates]
    result = sum(middle_pages)
    
    print(f"The sum of the middle pages from correctly ordered updates is: {result}")

# Specify the path to your input file
file_path = 'D:/Advent-of-code-2024/Advent-of-Code-2024/Day05/input.txt'
main(file_path)
